drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py

```diff
@@ -0,0 +1,590 @@
+        self.rot_bag = np.array([]).reshape(0, 3, 3)
+
+    def save_bag(self, req):
+        '''
+        Saves the buffers holding signals, and receiver array poses to a compressed .npz
+        file. These files can be used as source of input by running the passive sonar
+        driver with input_mode='bag' and providing filename in a rosparam.
+        '''
+        try:
+            np.savez_compressed(file=self.bag_filename, signal=self.signals_bag,
+                                trans=self.trans_bag, rot=self.rot_bag)
+
+        except IOError as e: # Couln't save to specified path
+            rospy.logerr(str(e))
+
+        finally:
+            self.bagging = False
+
+    def set_frequency(self, req):
+        '''
+        Sets the assumed frequency (in absence of noise) of the signals to received by the driver
+        '''
+        self.target_frequency = req.frequency
+        self.heading_start = np.empty((0, 3), float)
+        self.heading_end = np.empty((0, 3), float)
+        self.sample_variances = np.empty((0, 1), float)
+        self.pinger_position = np.array([np.NaN, np.NaN, np.NaN])
+        return {}
+
+    # Visualization
+
+    def visualize_dsp(self, t, signals, t_corr, cross_corr):
+        import matplotlib
+        matplotlib.use('agg')
+        import matplotlib.pyplot as plt
+        plt.plasma()
+
+        fig, axes = plt.subplots(nrows=2, ncols=1, sharex=False, sharey=False)
+        axes[0].set_title("Recorded Signals (Black is reference)")
+        axes[0].set_xlabel('Time (microseconds)')
+        axes[1].set_title("Cross-Correlations)")
+        axes[1].set_xlabel('Lag (microseconds)')
+
+        axes[0].plot(t, signals[0], color='black') # reference
+        axes[0].plot(t, signals[1:].T, )
+        axes[1].plot(t_corr, cross_corr.T)
+        fig.tight_layout(pad=2)
+        fig.canvas.draw() # render plot
+        plot_img = np.fromstring(fig.canvas.tostring_rgb(), # numpify
+                                 dtype=np.uint8, sep='')
+        plot_img = plot_img.reshape(fig.canvas.
+                                    get_width_height()[::-1] + (3,))
+        try:
+            self.plot_pub.publish(CvBridge().cv2_to_imgmsg(plot_img, 'bgr8'))
+        except CvBridgeError, e:
+            rospy.logerr(e)  # Intentionally absorb CvBridge Errors
+
+    def visualize_pinger_pos_estimate(self, bgra):
+        '''
+        Publishes a marker to RVIZ representing the last calculated estimate of the position of
+        the pinger.
+
+        rgba - list of 3 or 4 floats in the interval [0.0, 1.0] representing the desired color and
+            transparency of the marker
+        '''
+        marker = Marker()
+        marker.ns = "passive_sonar-{}".format(self.target_frequency)
+        marker.header.stamp = rospy.Time(0)
+        marker.header.frame_id = self.locating_frame
+        marker.type = marker.SPHERE
+        marker.action = marker.ADD
+        marker.scale.x = 0.2
+        np.clip(bgra, 0.0, 1.0)
```

Comment:
This line doesn't mutate the `bgra` values, does it? `bgra = np.clip(bgra, ...)`